server.js

```diff
@@ -1066,6 +1066,81 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+// Liberapay integration.
+camp.route(/^\/liberapay\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var entity = match[1];  // eg, 'Changaco'
+  var type = match[2];
+  var format = match[3];
+  var apiUrl = 'https://liberapay.com/' + entity + '/public.json';
+  // Establish default
+  var badgeData = getBadgeData('receives', data);
+  switch(type) {
+      case 'gives':
+          badgeData = getBadgeData('gives', data);
+          break;
+      case 'patrons':
+          badgeData = getBadgeData('patrons', data);
+          break;
+      case 'goal':
+          badgeData = getBadgeData('goal progress', data);
+          break;
+      }
+  if (badgeData.template === 'social') {
+    badgeData.logo = getLogo('liberapay', data);
+  }
+  request(apiUrl, function dealWithData(err, res, buffer) {
+    if (err != null) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+    try {
+      var data = JSON.parse(buffer);
+      var value;
+      var currency;
+      switch(type) {
+        case 'receives':
+            if (data.receiving) {
+                value = data.receiving.amount;
+                currency = data.receiving.currency;
+                badgeData.text[1] = metric(value) + ' ' + currency + '/week';
+                }
+            break;
+        case 'gives':
+            if (data.giving) {
+                value = data.giving.amount;
+                currency = data.giving.currency;
+                badgeData.text[1] = metric(value) + ' ' + currency + '/week';
+                }
+            break;
+        case 'patrons':
+            value = data.npatrons;
+            badgeData.text[1] = metric(value);
+            break;
+        case 'goal':
+            if (data.goal) {
+                value = Math.round(data.receiving.amount/data.goal.amount*100);
+                badgeData.text[1] = metric(value) + '%';
```

Comment:
Probably this should not use `metric` since it is a percentage.

Also, could you check `Number.isNaN(value)` to handle the division by zero case, so we don't put `NaN%` on a badge?

Reply:
Is this really necessary? You can't set a zero goal on the Liberapay side.

Reply:
It's worth noting this isn't a hypothetical. I've no idea what this API might return when things go wrong. **build | undefined** and **coverage | NaN%** have shown up as example badges on shields.io. There's even a **NaN%** up there right now that needs fixing.

The approach I'd like to take in the future is using a schema to validate the important aspects of the response. That way we don't have to write a lot of defensive code, and we just get good error messages automatically. That'll happen in #963 and beyond, so no need to fuss about this now.